ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/unixext.ml

```diff
@@ -586,6 +585,55 @@ let with_polly_wait kind fd f =
       in
       f wait fd
 
+module Timeout = struct
+  type t = Mtime.Span.t
+
+  let of_span s : t = s
+
+  let of_string str =
+    let seconds = Float.of_string str in
+    match Mtime.Span.of_float_ns (seconds *. 1e9 |> Float.round) with
+    | Some span ->
+        span
+    | None ->
+        (* e.g. negative, or >=2^53. *)
+        invalid_arg (Printf.sprintf "Span out of range: %gs" seconds)
+
+  (* nanosecond conversion doesn't roundtrip, so use only microseconds, which does *)
+  let to_string t = Mtime.Span.to_float_ns t *. 1e-9 |> Printf.sprintf "%.6f"
+
+  let pp = Mtime.Span.pp
+end
+
+module Timer = struct
+  let delay dt = Mtime.Span.to_float_ns dt *. 1e-9 |> Thread.delay
```

Comment:
What is a timer? Some comment that introduces the abstraction.

Reply:
I'm adding some ocamldoc comments. Here is what I have so far:
```
+(** A "timeout" is a time span, relative to the start of an operation.
+  It is usually used to express a limit on the duration of an operation.
+
+  Not all time spans are timeouts, e.g. the difference between 2 timestamps (a duration) is not a timeout,
+  because it only expresses measured, elapsed time, not a limit.
+
+  All timeouts are time spans, e.g. a 30s timeout on a read operation means that we should raise an exception if the
+  read operation doesn't complete within 30s from the moment it started.
+ *)
 module Timeout : sig
   type t = private Mtime.Span.t
+  (** upper bound for the duration of an operation *)
```

```
+(** A timer is started at the same time as a particular operation.
+  It measures elapsed time from that moment, and has a desired upper bound, its timeout.
+
+  At any moment we can query the timer and get the remaining time, which is either
+   [Spare] if the timeout has not been reached yet,
+   or [Excess] if the timer has already expired, i.e. we have already exceeded the timeout.
+
+ *)
 module Timer : sig
   val delay : Mtime.Span.t -> unit
   (** [delay dt] delays the execution of the current thread for [dt]. *)

   type t
+  (** a timer that was started at a particular moment in time, and measures time relative to it *)

   type remaining = Spare of Mtime.Span.t | Excess of Mtime.Span.t

   val start : timeout:Timeout.t -> t
-  (** [start ~timeout] *)
+  (** [start ~timeout] starts a timer, with the relative deadline [timeout]. *)
```

Perhaps we should call 'Excess' as 'Expired' instead, it'd fit with the typical usage of a timer better.

Reply:
We are now using the Timer module from Clock which has equivalent documentation comments.